docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,492 @@
+- Where necessary, we will import behaviours from the MachineSet and MachineDeployment controllers.
+- The controller should tolerate the manual removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+
+##### Control plane  healthcheck
+
+- Will be used during scaling and upgrade operations.
+- There will also be a signal to indicate a healthcheck that could not be made, e.g. due to a network partition.
+- For stacked control planes, we will present etcd quorum status within the Status.Ready field, and also report the number of active cluster members through "ReadyReplicas"
+- The addition of status fields will be done conservatively to prevent them being relied upon even as we choose to deprecate them in the future.
+- Unlike previous lifecycle managers (see Alternatives), ComponentStatus will be not be used as it is deprecated.
+- Examples of specific checks we may perform include:
+  - Checking the managed cluster has a kubeadm-config configmap, which will mark the cluster as Initialized
+  - Checking the managed cluster has a kubelet-$version-configmap matching the desired
+     version to check the state of upgrades
+  - Checking the status of all pods marked with the system-cluster-critical priorityClass within the managed cluster, and providing a count of ready vs total.
+  - Running PodExec etcdctl, or port-forwarding to etcd to get etcd cluster health information
+  - Checking the status of pods marked with the app=kube-dns label as a proxy for information about the health of CNI.
+
+##### Adoption of pre-v1alpha3 Control Plane Machines
+
+- Existing Control Plane Machines will need to be updated with labels matching the expected label selector.
+- The KubeadmConfigSpec can be re-created from the referenced KubeadmConfigs for the Machines matching the label selector.
+  - If there is not an existing initConfiguration/clusterConfiguration only the joinConfiguration will be populated.
+- In v1alpha2, the Cluster API Bootstrap Provider is responsible for generating certificates based upon the first machine to join a cluster. The OwnerRef for these certificates are set to that of the initial machine, which causes an issue if that machine is later deleted. For v1alpha3, control plane certificate generation will be replicated in the ControlPlane provider. Given that for v1alpha2 these certificates are generated with deterministic names, i.e. prefixed with the cluster name, the migration mechanism should replace the owner reference of these certificates during migration. The bootstrap provider will need to be updated to only fallback to the v1alpha2 secret generation behavior if Cluster.Spec.ControlPlaneRef is nil.
```

Comment:
> control plane certificate generation will be replicated in the ControlPlane provider
This should be outlined in the the behavioral changes section above, it wasn't clear while reading top to bottom